proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,148 @@
+---
+simd: '0023'
+title: Asynchronous Program Execution and Broadcast
+authors:
+  - Anatoly Yakovenko
+category: Standard/Meta
+type: Core
+status: Draft
+created: 2024-04-29
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This feature changes how the ledger is broadcast and executed. It
+separates proposing blocks full of user transactions from blocks
+with votes.  It allows for N concurrent builders of user transaction
+blocks, and it allows for asynchronous execution of user blocks.
+
+## Motivation
+
+1. Single leader for user transactions is a bottleneck.  Leaders
+are in a single spot in the world, while clients are all over the
+world. Latency is therefore based on how close a client is to a
+leader.  Leaders also have to spend a ton of resources on prioritization
+of transactions.
+
+2. Executing programs before voting is a bottleneck. Fork choice
+doesn't depend on program execution, and delaying voting creates
+forks and also delays the next block producer from starting.
+
+## Alternatives Considered
+
+APEX design that was proposed in https://github.com/solana-labs/solana/pull/24127
+
+## New Terminology
+
+* UserBlock - a block full of none vote transactions.
+
+* UserBlockEntry - a hash of the block that was proposed
+
+* UserBlockSlot - a slot measuring half a normal slot in ticks.
+
+* Leader - the current leader for the slot that will propose a PoH
+ledger full of Votes and UserBlockEntry
+
+* Builder - a node that is scheduled to propose a block with non
+vote transactions
+
+## Detailed Design
+
+### Overview
+Leaders are scheduled to propose blocks as they are currently by
+the LeaderSchedule.
+
+Builder's are scheduled along side leaders to propose blocks. N
+number of builders can be scheudled per slot.  Builder are
+scheudled at 2x the rate of leaders.  So for every leader slot,
+there are N builders producing 2 UserBlocks.
+
+While a leader is scheduled, they receive and encode votes as normal.
+Any well formed UserBlocks that were received from the previous
+UserBlockSlot are added to the leaders PoH ledger as UserBlockEntry.
+
+The N concurrent Builder have 200ms slots to create blocks out
+of user transactions. These are transmitted to the network.  The
+leader receives and decodes them and generates a UserBlockEntry,
+and adds it to PoH as soon as the leaders PoH has passed the
+UserBlockSlot.
+
+### Fork Choice
+
+If a validator doesn't have the builder's UserBlock, the validator
```

Comment:
im confused where they repair from here if they're the only one getting the UserBlock. unless its propagated to the entire cluster somehow

Reply:
UserBlocks are sent over turbine. So repair and everything else works as is.

Reply:
That is correct.  Before voting on a fork each validator has to have all the data, including all the data form all the user blocks in that fork.  Otherwise there is no way to guarantee that everyone can execute because the data could be withheld